rfcs/20221020-tensor-indexed-updates.md

```diff
@@ -0,0 +1,468 @@
+indices = tf.transpose(tf.stack([dim0, dim1, dim2]))
+
+# Create a set of values with the same number of elements in indices, since
+# tensor_scatter_* currently doesn't support broadcasting.
+updates = tf.broadcast_to(tf.constant(0, dtype=tensor.dtype),dim_shape)
+
+# Scatter updates into x.
+x = tf.tensor_scatter_nd_update(x, indices, updates)
+```
+
+**Setting negative values to zero:**
+```python
+#========================================================================
+# Numpy:
+#========================================================================
+x[x < 0] = 0
+
+#========================================================================
+# TensorFlow via tf.tensor_scatter_nd_update:
+#========================================================================
+# Extract index tensor from boolean mask.
+indices = tf.where(x < 0)
+# Build values tensor.
+dim_size = tf.shape(indices)[0:1]
+values = tf.broadcast_to(tf.constant(0, dtype=x.dtype), dim_size)
+# Scatter updates into x.
+x = tf.tensor_scatter_nd_update(x, indices, values)
+
+#========================================================================
+# TensorFlow via multiplexing tf.where:
+#========================================================================
+# NOTE: this is only equivalent to NumPy for simple cases where the "updates"
+# is a scalar.  Otherwise, the RHS to `x[mask] = RHS` needs to match the shape
+# of indices returned via `tf.where(mask)`, which is not the same shape as the
+# mask itself.
+x = tf.where(x < 0, 0, x)
+```
+
+As you can see, although it still is technically possible to perform many of the
+same operations with using TensorFlow's existing functionality, it's quite
+cumbersome.
+
+## Proposed Syntax
+
+Rather than invent a new API syntax, we will rely on the successful rollout of
+indexed updates in JAX - which also has the immutable tensor restriction - via
+[`.at[]`](https://jax.readthedocs.io/en/latest/_autosummary/jax.numpy.ndarray.at.html#jax.numpy.ndarray.at).
+
+
+```python
+#========================================================================
+# Proposed TensorFlow Syntax:
+#========================================================================
+# Setting a tensor slice to zero:
+x = x.at[:,:,2].set(0)
+
+# Setting negative values to zero:
+x = x.at[x < 0].set(0)
+```
+
+For `tf.Variable`, we have already had _some_ support for in-place slice updates
+via the `.assign()` method:
+```python
+# Existing tf.Variable in-place slice assignment:
+v[:, :, 2].assign(0)
+```
+For consistency with NumPy, we propose to introduce direct assignment as well:
+```python
+# Proposed tf.Variable in-place slice assignment:
+v[:, :, 2] = 0
+```
+This would modify the `tf.Variable` in-place.  For consistency with the proposed
+`Tensor` API, we would also allow
+```python
+# Proposed tf.Variable copy-and-update assignment:
+v = v.at[:, :, 2].set(0)
+```
+Which would create a modified copy of the original variable.
+
```

Comment:
A possible naive confusion is the user thinking that `at[]` by itself returns something usable, a mutable Tensor constructed from the sliced values. In reality, the user only gets something usable (outside of the indexed update operation) by "finishing the sentence" and calling `set()`.

This could perhaps be mitigated with clear documentation and informative error messages.

Reply:
Yes.  Hopefully through usage examples and the tutorial, it would become clear.  I'm not sure if there's a python equivalent to `ABSL_MUST_USE_RESULT` to force the user to actually assign the result of something like `set(...)` to something, which may help mitigate this.